File: src/main/java/essentials/feature/Exp.java
package essentials.feature;

import essentials.core.player.PlayerData;
import essentials.internal.Bundle;
import mindustry.entities.type.Player;
import mindustry.gen.Call;

import static essentials.Main.config;
import static essentials.Main.playerDB;

public class Exp {
    final double BASE_XP = config.basexp();
    final double EXPONENT = config.exponent();

    public Exp(Player player) {
        PlayerData target = playerDB.get(player.uuid);

        int currentlevel = target.level();
        int max = (int) calculateFullTargetXp(currentlevel);

        int xp = target.exp();
        int levelXp = max - xp;
        int level = calculateLevel(xp);
        int reqexp = (int) Math.floor(max);
        String reqtotalexp = xp + "(" + (int) Math.floor(levelXp) + ") / " + (int) Math.floor(max);

        target.reqexp(reqexp);
        target.level(level);
        target.reqtotalexp(reqtotalexp);

        if (currentlevel < level && currentlevel > config.alarmlevel() && config.levelupalarm())
            Call.onInfoToast(new Bundle(target.locale()).get("player.levelup", player.name, level), 600);
    }

    double calcXpForLevel(int level) {
        return BASE_XP + (BASE_XP * Math.pow(level, EXPONENT));
    }

    double calculateFullTargetXp(int level) {
        double requiredXP = 0;
        for (int i = 0; i <= level; i++) {
            requiredXP += calcXpForLevel(i);
        }
        return requiredXP;
    }

    int calculateLevel(double xp) {
        int level = 0;
        double maxXp = calcXpForLevel(0);
        do {
            maxXp += calcXpForLevel(++level);
        } while (maxXp < xp);
        return level;
    }
}

File: src/main/kotlin/essentials/features/Exp.kt
package essentials.features

import essentials.Config
import essentials.Main
import essentials.PlayerData
import essentials.internal.Bundle
import mindustry.gen.Call
import kotlin.math.floor
import kotlin.math.pow

class Exp(target: PlayerData) {
    val baseXP = Config.baseXp
    val exponent = Config.exponent
    private fun calcXpForLevel(level: Int): Double {
        return baseXP + baseXP * level.toDouble().pow(exponent)
    }

    private fun calculateFullTargetXp(level: Int): Double {
        var requiredXP = 0.0
        for (i in 0..level) {
            requiredXP += calcXpForLevel(i)
        }
        return requiredXP
    }

    private fun calculateLevel(xp: Double): Int {
        var level = 0
        var maxXp = calcXpForLevel(0)
        do {
            maxXp += calcXpForLevel(++level)
        } while (maxXp < xp)
        return level
    }

    init {
        val currentlevel = target.level
        val max = calculateFullTargetXp(currentlevel).toInt()
        val xp = target.exp
        val levelXp = max - xp
        val level = calculateLevel(xp.toDouble())
        val reqexp = floor(max.toDouble()).toInt()
        val reqtotalexp = xp.toString() + "(" + floor(levelXp.toDouble()).toInt() + ") / " + floor(max.toDouble()).toInt()
        target.reqexp = reqexp
        target.level = level
        target.reqtotalexp = reqtotalexp
        if (currentlevel < level && currentlevel > Config.alarmLevel && Config.levelUpAlarm) Call.infoToast(Bundle(target.locale)["player.levelup", target.name, level], 600f)
    }
}